### ISSUE TITLE
Emergency Mode Power off RemoteNFCpanel

### ISSUE THREAD

**Issue by WhiskeyGolf** [FrSkyRC/ETHOS-Feedback-Community#4873]
When testing the emergency mode on a model with an RemoteNFCpanel on PowerSwitch, I noticed that the model simply switches off in emergency mode. Should that be the case? It doesn't really feel like flight-safe.

**Comment by WhiskeyGolf** [comment]
Unfortunately, it is not entirely clear to me what still works in the event of emergency mode. After testing my models, it is clear that the probability of loss is very high. Unfortunately, in the year since I started using Ethos, it has been in emergency mode several times. I have never had such a mode with any other system. It is starting to make me a little suspicious. I hope there is a solution here soon. Thanks

**Comment by dean5378** [comment]
My TD X20R went into EMERGENCY mode all of a sudden and I had to reboot the TX. What the heck happened? I have never had a TX do that before.  If I was in the air it would have been disastrous!

**Comment by robthomson** [comment]
What firmware you running?

**Comment by mgsanna** [comment]
> My TD X20R went into EMERGENCY mode all of a sudden and I had to reboot the TX. What the heck happened? I have never had a TX do that before. If I was in the air it would have been disastrous!

If this happen in Air, you can still control the plane without any issue, of course you have to land as soon as possible, but the rf section and basic control of 4 axis is always active... so no disaster, lot of time emergency mode is "detonated" from corrupted SD card or something that can "alert" watchdog of the system during writing memory or something similar, sometime there is a corrupted file, installed by user just like sound file or similar
and as @robthomson  ask you what firmware and what was the circumstance... of your "emergency mode"

**Comment by mgsanna** [comment]
> Unfortunately, it is not entirely clear to me what still works in the event of emergency mode. After testing my models, it is clear that the probability of loss is very high. Unfortunately, in the year since I started using Ethos, it has been in emergency mode several times. I have never had such a mode with any other system. It is starting to make me a little suspicious. I hope there is a solution here soon. Thanks

Strange, I'm using Ethos since the origins.... spring 2021 and had only a "emergency mode"  during a flight of one of my jet, it was caused by a defective SD-Card, nothing was lost... 
And of course some times "emergency mode" is caused by some "memory bug" or some lua widget

**Comment by robthomson** [comment]
Only time you will see these is.

- running a beta firmware (expect a bug)
- a physical issue like a faulty sdcard.

**Comment by dean5378** [comment]
I am running the latest released version of the firmware.
Does the radio come with a SD card?  I did not install one... I thought files were being saved within the internal memory???

**Comment by dean5378** [comment]
@robthomson @mgsanna 
I am running the latest released version of the firmware.
Does the radio come with a SD card? I did not install one... I thought files were being saved within the internal memory???

**Comment by dean5378** [comment]
I had been working on programming a new plane.  I was just holding the TX when it went into Emergency Mode.  I do not remember if the receiver was on or off..

**Comment by robthomson** [comment]
As in 1.5.19?   Or RC8

**Comment by mgsanna** [comment]
> @robthomson @mgsanna I am running the latest released version of the firmware. Does the radio come with a SD card? I did not install one... I thought files were being saved within the internal memory???

X20RS come with NAND internal memory, so can you tell what Firmware are you on (you can see it on Info in system page), what TD ISRM module firmware version, what  receiver if was bound to a receiver, and what were you "doing" just before "emergency"?
it's to better understand what happen in your case...
(just for your info there is also a widget called Emergency mode" to test and invoke an emergency and test functionally your model

**Comment by dean5378** [comment]
@robthomson @mgsanna  Running 1.5.19,FCC onTX
Receiver AP_SR10+_R10+_1.0.11.frsk 
Firmware Version【v1.0.11】
Not sure the receiver was on/or not but I was just holding the TX in my hand.  Screen went to Emergency Screen until I rebooted.

**Comment by robthomson** [comment]
Running any lua scripts?

**Comment by dean5378** [comment]
@robthomson No...I can not remember if I had just run the STAB Cali or not.. I remember just holding the TX in my left hand.

**Comment by dean5378** [comment]
@robthomson How does one start the "Emergency Test Widget?"

**Comment by dean5378** [comment]
@robthomson The other thing that I may have been doing was "playing with" the TX gyro in a Mix to see how values change when the TX is moved..this may be why I was holding the TX up and moving it.

**Comment by mgsanna** [comment]
> @robthomson The other thing that I may have been doing was "playing with" the TX gyro in a Mix to see how values change when the TX is moved..this may be why I was holding the TX up and moving it.

I'don't think gyro can induce Emergency mode and also movement is not the culprit

**Comment by dean5378** [comment]
@robthomson Is it safe to use my TX?

**Comment by mgsanna** [comment]
> @robthomson How does one start the "Emergency Test Widget?"

You have to download it from github and install in the scripts folder of your radio
(sorry I've edited your post by mistake, I was answering you... sorry)

**Comment by mgsanna** [comment]
> @robthomson Is it safe to use my TX?

Can you reproduce "emergency mode"? if yes I'll try to understand what is the cause
if not you can use your radio without fear, because as stated before, you still have the control of your model...(4 principal axis)

**Comment by mgsanna** [comment]
@dean5378 if you want to be more "sure" make a backup and start the model from scratch....

**Comment by robthomson** [comment]
It's possible you have hit a bug somewhere.

Need to just try narrow down / make it repeatable.

**Comment by strgaltdel** [comment]
emergency modes are very rare and usually have a systematic cause.
if you have this effect so often, it is likely to have a rather atypical background

to completely rule out the possibility of badly programmed luas compromising the system, i would rename the scripts folder, then the lua point can definitely not play a role

**Comment by dean5378** [comment]
@robthomson @mgsanna Ok guys, thanks for all of the input!  If it happens again I will take stringent notes and let you know.

**Comment by bsongis** [comment]
The original issue is still very different from this discussion, and perhaps it is repeatable @WhiskeyGolf ? The NFC Switch will perhaps not like the loss of RF during the time of the Watchdog Reset (Emergency Mode script)?

**Comment by bsongis** [comment]
@dean5378 I will really appreciate if you succeed to reproduce the problem! Emergency Mode should not happen. It's there to force the radio to reboot in case something bad happens (software bug in most cases). Lua scripts can be used in so many ways, it is difficult to ensure that everything bad is catched by the OS! The gyro ... I doubt, but who knows ... Any information will be useful. But perhaps in another issue to avoid mixing the topics?

**Comment by WhiskeyGolf** [comment]
Hello @bsongis , thanks for the answer! 
It can be repeated. 
The switch is bind to channel 15. What position is channel 15 in emergency mode? Up to which channel can I still control? 
In several models my flaps go to the middle position in the EM and back and then I can no longer control them. I can also no longer move the gear in the EM either.

**Comment by bsongis** [comment]
Would you please attach your model file? I need to check, it all depends on the behaviour of channel 15 on startup!
In emergency mode, all your mixes / switches / RF will run as usual. But there is a small delay (during the reboot of the transmitter) . And I am afraid that it is during this delay that the switch will shutdown. Also did you configure the Failsafe on this channel? It should remain ON of course!

**Comment by WhiskeyGolf** [comment]
[l-39_gfkturbtw.bin.zip](https://github.com/user-attachments/files/18141182/l-39_gfkturbtw.bin.zip)

**Comment by WhiskeyGolf** [comment]
But in EM I can't even extend the gear. 
The failsafe for CH15 is on hold. 
In the failsafe test the NFC stays on, so it works.

**Comment by WhiskeyGolf** [comment]
The model in the attached file is also the model in which the EM occurred during a screenshot in another issue.

**Comment by bsongis-frsky** [comment]
Trying ...

**Comment by bsongis-frsky** [comment]
@WhiskeyGolf which is the Ethos version you are using?

**Comment by WhiskeyGolf** [comment]
Ethos 1.5.19
and latest RemoteNFC Version

**Comment by bsongis** [comment]
I think that the way that the NFC switch is programmed is dangerous:
- Your CH15 (NFC Switch) is controlled by the LS "SYSTEM"
- "SYSTEM" is OFF on startup (which is the case on EM)

So it works like it should. 
But it is very dangerous, in fact you would like that the logic switch keeps ON after the radio reboots!

**Comment by bsongis** [comment]
![image](https://github.com/user-attachments/assets/a239ed45-0632-46f2-9f75-4526684be742)

**Comment by WhiskeyGolf** [comment]
Yes, I used a logic switch to prevent a single switch from being turned off. However, when I turn on my RX, the switch for CH15 is always off. I don't want the model to turn on immediately. What would be a good idea to program CH15? I think, as mentioned above, it needs to give you the EM a little more information.

**Comment by bsongis** [comment]
Yes I do understand your logic, which seems good to me. But it remains that on startup your logic switch "SYSTEM" is always OFF, even when rebooting from emergency mode. I need to find a solution for that. Obviously in such case, the state of the switch should be kept, but it is not a simple change that I can do just before a release!

**Comment by Toffs** [comment]
Hi,
In my view, the issue is the default programming for NFC - see this thread:
https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/3934

The NFC Remote ‘off’ should be a positive/intentional channel change. Currently if anything at all interrupts the +100 channel output then the NFC switches off.
This is backwards, it should be the default output (which is 0) is ‘on’ and the off function should require the +100 on a channel.

That users are required to program this functionality themselves is not right.

@bsongis-frsky - I’m really happy to discuss / try to explain this clearer if needed (DMs or Email).

**Comment by WhiskeyGolf** [comment]
This is a very good idea, and would certainly not be a major update on the hardware side

**Comment by bsongis-frsky** [comment]
As written in #3934 the idea proposed by @Toffs seems to me very good, and fixes the design issue explained above.

I have pushed it this morning to FrSky and it seems accepted. Good news should come soon!

**Comment by bsongis-frsky** [comment]
There is nothing else to do in Ethos for this.

**Comment by pstasek** [comment]
When implementing the NFC switch, the behavior in Emergency Mode or any other non-standard scenario was one of my main concerns.

My programming was done to default at Power On state, meaning reversed logic.  I configured a logical switch called Power Off which is active only when a certain condition is met, i.e.: Ignition is OFF and a sticky LSW is on which is triggered by a momentary switch. This means, that even if the radio should power cycle mid-flight for some reason (intermittent radio battery connection for example), the switch always defaults to ON. 
I also find it practical if the model is powered up when the radio is switched on. If I have multiple models on the field, I switch it off and then swap a model. If the correct model is not selected, the NFC switch remains off obviously. Think about it @WhiskeyGolf 

Now to the point... when programmed like this, it also successfully passes the Emergency mode test using the EM Test Lua script (see the video below). It is handy for testing such a scenario and should be part of the setup process for every complex model.

https://photos.app.goo.gl/W4374gYXEvv9F9RN7

**Comment by WhiskeyGolf** [comment]
I secured the NFC via the Engine On Switch using a replace mix. Now it stays on in all situations. It's already been tested. But I still think the idea @Toffs  is much better.

**Comment by EngelMT** [comment]
This must be changed via the hardware/software of the NFC. An electronic switch, regardless of how it is operated, must always be on. This was also wrong with the first RBs. This was then changed on my advice. It must basically be a "closed" contact that is switched off by an "OFF" pulse. If there is no pulse, it must automatically be "ON" (closed).

**Comment by Toffs** [comment]
Yes, this is safest.
However, might I suggest having the option for resume 'last state' as well.  Maybe that won't work as it might need a continuous power feed to save this state, like how computer BIOS allows this option because it uses the small button cell power feed when the mains is off.
If it it is possible to program, then having the option for either 'always on' or 'last state' would be good.

Remember, we're trying to mitigate against all unknown possibilities that might cause the state of the NFC to change unexpectedly.  RF interference, incorrect Ethos Programming, unanticipated transmitter behaviour or state (emergency mode).

Some people might want to keep the current default behaviour of 'no signal = off' - so perhaps having this as an option as well might be good to keep current setups happy/safe?

**Comment by pstasek** [comment]
Don't forget, the range of the remote switch is rather short, so it already remains in the last state under no-signal conditions until the receiver or RB unit is power cycled. In case of a power cycle of the model (intermittent battery), it always defaults to ON which is the only safe and correct option.

"No signal = Off" behaviour would be very dangerous as it would create a hazard in case of intermittent battery or voltage drop in the system.

The model should deal with the no-signal scenario according to the fail-safe settings in the Power-on state. The Ethos programming, fail-safe configuration and testing of the Emergency mode should remain the responsibility of the user. 

Based on my ground testing, the Remote NFC panel works exactly as it should. When I turn off the radio, the model power remains in the last state, be it on or off. When I trigger the Emergency mode, I can still control the plane as usual (see the video in my previous post). If I plug in the model battery, the model comes alive, no matter if the radio is on or off. 

As a "nice to have" on the Ethos side, I can imagine a dedicated NFC Remote switch mix with features tailored for this feature with a simple setup, such as choosing the switch, active condition (like ignition off or throttle cut active), and output channel and it would take care of all the necessary functionality such as confirmation dialogue, ensuring the ON state in E.M., etc...

**Comment by WhiskeyGolf** [comment]
” As a "nice to have" on the Ethos side, I can imagine a dedicated NFC Remote switch mix with features tailored for this feature with a simple setup, such as choosing the switch, active condition (like ignition off or throttle cut active), and output channel and it would take care of all the necessary functionality such as confirmation dialogue, ensuring the ON state in E.M., etc... “ 

That's actually a very good idea that I've already thought of!

**Comment by Dsrenger** [comment]
I just saw an Firmware update for the RemoteNFC switch

![Image](https://github.com/user-attachments/assets/d15d14e6-c338-4067-b8e6-17eb91c6fcda)

Is it linked to this issue?
What does "1. Optimized the control logic of the switch channel." mean?

Do we have to change something in the way we handle?

Sometimes changenotes should be a little bit longer than 1 sentance.

Can someone from FrSky elaborate?

Regards
Dennis

**Comment by Dsrenger** [comment]
Any info on this one?

The patch note is not telling me a lot.

**Comment by Dsrenger** [comment]
No one?

**Comment by Toffs** [comment]
Hi @Dsrenger 

Been a bit slow testing this, but update one of my NFC's to the latest firmware.  See my comment on the related issue about what this update does: [issue3934](https://github.com/FrSkyRC/ETHOS-Feedback-Community/issues/3934)

In my opinion, this means the NFC Remote Panel now has proper failsafe design and a good trigger solution.  I'm now installing the devices in my main models :)

**Comment by Dsrenger** [comment]
Nice, thanks for your test.

As per the patch notes, you have to upgrade to 1.6 and therefore to HF 3.0.x

Don't plan to update for now.

Thanks for your test. At least we know what the update is doing.
